File: src/colmap/feature/types_test.cc
// Copyright (c), ETH Zurich and UNC Chapel Hill.
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
//     * Redistributions of source code must retain the above copyright
//       notice, this list of conditions and the following disclaimer.
//
//     * Redistributions in binary form must reproduce the above copyright
//       notice, this list of conditions and the following disclaimer in the
//       documentation and/or other materials provided with the distribution.
//
//     * Neither the name of ETH Zurich and UNC Chapel Hill nor the names of
//       its contributors may be used to endorse or promote products derived
//       from this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE
// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.

#include "colmap/feature/types.h"

#include "colmap/math/math.h"

#include <gtest/gtest.h>

namespace colmap {
namespace {

TEST(FeatureKeypoints, Nominal) {
  FeatureKeypoint keypoint;
  EXPECT_EQ(keypoint.x, 0.0f);
  EXPECT_EQ(keypoint.y, 0.0f);
  EXPECT_EQ(keypoint.a11, 1.0f);
  EXPECT_EQ(keypoint.a12, 0.0f);
  EXPECT_EQ(keypoint.a21, 0.0f);
  EXPECT_EQ(keypoint.a22, 1.0f);

  FeatureKeypoints keypoints(1);
  EXPECT_EQ(keypoints.size(), 1);
  EXPECT_EQ(keypoints[0].x, 0.0f);
  EXPECT_EQ(keypoints[0].y, 0.0f);
  EXPECT_EQ(keypoints[0].a11, 1.0f);
  EXPECT_EQ(keypoints[0].a12, 0.0f);
  EXPECT_EQ(keypoints[0].a21, 0.0f);
  EXPECT_EQ(keypoints[0].a22, 1.0f);

  keypoint = FeatureKeypoint(1, 2);
  EXPECT_EQ(keypoint.x, 1.0f);
  EXPECT_EQ(keypoint.y, 2.0f);
  EXPECT_EQ(keypoint.a11, 1.0f);
  EXPECT_EQ(keypoint.a12, 0.0f);
  EXPECT_EQ(keypoint.a21, 0.0f);
  EXPECT_EQ(keypoint.a22, 1.0f);
  EXPECT_NEAR(keypoint.ComputeScale(), 1.0f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeScaleX(), 1.0f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeScaleY(), 1.0f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeOrientation(), 0.0f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeShear(), 0.0f, 1e-6);

  keypoint = FeatureKeypoint(1, 2, 0, 0);
  EXPECT_EQ(keypoint.x, 1.0f);
  EXPECT_EQ(keypoint.y, 2.0f);
  EXPECT_EQ(keypoint.a11, 0.0f);
  EXPECT_EQ(keypoint.a12, 0.0f);
  EXPECT_EQ(keypoint.a21, 0.0f);
  EXPECT_EQ(keypoint.a22, 0.0f);

  keypoint = FeatureKeypoint(1, 2, 1, 0);
  EXPECT_EQ(keypoint.x, 1.0f);
  EXPECT_EQ(keypoint.y, 2.0f);
  EXPECT_EQ(keypoint.a11, 1.0f);
  EXPECT_EQ(keypoint.a12, 0.0f);
  EXPECT_EQ(keypoint.a21, 0.0f);
  EXPECT_EQ(keypoint.a22, 1.0f);
  EXPECT_NEAR(keypoint.ComputeScale(), 1.0f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeScaleX(), 1.0f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeScaleY(), 1.0f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeOrientation(), 0.0f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeShear(), 0.0f, 1e-6);

  keypoint = FeatureKeypoint(1, 2, 1, M_PI / 2);
  EXPECT_EQ(keypoint.x, 1.0f);
  EXPECT_EQ(keypoint.y, 2.0f);
  EXPECT_NEAR(keypoint.a11, 0.0f, 1e-6);
  EXPECT_NEAR(keypoint.a12, -1.0f, 1e-6);
  EXPECT_NEAR(keypoint.a21, 1.0f, 1e-6);
  EXPECT_NEAR(keypoint.a22, 0.0f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeScale(), 1.0f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeScaleX(), 1.0f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeScaleY(), 1.0f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeOrientation(), M_PI / 2, 1e-6);
  EXPECT_NEAR(keypoint.ComputeShear(), 0.0f, 1e-6);

  keypoint = FeatureKeypoint(1, 2, 2, M_PI / 2);
  EXPECT_EQ(keypoint.x, 1.0f);
  EXPECT_EQ(keypoint.y, 2.0f);
  EXPECT_NEAR(keypoint.a11, 0.0f, 1e-6);
  EXPECT_NEAR(keypoint.a12, -2.0f, 1e-6);
  EXPECT_NEAR(keypoint.a21, 2.0f, 1e-6);
  EXPECT_NEAR(keypoint.a22, 0.0f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeScale(), 2.0f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeScaleX(), 2.0f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeScaleY(), 2.0f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeOrientation(), M_PI / 2, 1e-6);
  EXPECT_NEAR(keypoint.ComputeShear(), 0.0f, 1e-6);

  keypoint = FeatureKeypoint(1, 2, 2, M_PI);
  EXPECT_EQ(keypoint.x, 1.0f);
  EXPECT_EQ(keypoint.y, 2.0f);
  EXPECT_NEAR(keypoint.a11, -2.0f, 1e-6);
  EXPECT_NEAR(keypoint.a12, 0.0f, 1e-6);
  EXPECT_NEAR(keypoint.a21, 0.0, 1e-6);
  EXPECT_NEAR(keypoint.a22, -2.0f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeScale(), 2.0f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeScaleX(), 2.0f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeScaleY(), 2.0f, 1e-6);
  EXPECT_TRUE(std::abs(keypoint.ComputeOrientation() - M_PI) < 1e-6 ||
              std::abs(keypoint.ComputeOrientation() + M_PI) < 1e-6);
  EXPECT_NEAR(keypoint.ComputeShear(), 0.0f, 1e-6);

  keypoint = FeatureKeypoint::FromShapeParameters(1, 2, 2, 2, M_PI, 0);
  EXPECT_EQ(keypoint.x, 1.0f);
  EXPECT_EQ(keypoint.y, 2.0f);
  EXPECT_NEAR(keypoint.a11, -2.0f, 1e-6);
  EXPECT_NEAR(keypoint.a12, 0.0f, 1e-6);
  EXPECT_NEAR(keypoint.a21, 0.0, 1e-6);
  EXPECT_NEAR(keypoint.a22, -2.0f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeScale(), 2.0f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeScaleX(), 2.0f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeScaleY(), 2.0f, 1e-6);
  EXPECT_TRUE(std::abs(keypoint.ComputeOrientation() - M_PI) < 1e-6 ||
              std::abs(keypoint.ComputeOrientation() + M_PI) < 1e-6);
  EXPECT_NEAR(keypoint.ComputeShear(), 0.0f, 1e-6);

  keypoint = FeatureKeypoint::FromShapeParameters(1, 2, 2, 3, M_PI, 0);
  EXPECT_EQ(keypoint.x, 1.0f);
  EXPECT_EQ(keypoint.y, 2.0f);
  EXPECT_NEAR(keypoint.a11, -2.0f, 1e-6);
  EXPECT_NEAR(keypoint.a12, 0.0f, 1e-6);
  EXPECT_NEAR(keypoint.a21, 0.0, 1e-6);
  EXPECT_NEAR(keypoint.a22, -3.0f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeScale(), 2.5f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeScaleX(), 2.0f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeScaleY(), 3.0f, 1e-6);
  EXPECT_TRUE(std::abs(keypoint.ComputeOrientation() - M_PI) < 1e-6 ||
              std::abs(keypoint.ComputeOrientation() + M_PI) < 1e-6);
  EXPECT_NEAR(keypoint.ComputeShear(), 0.0f, 1e-6);

  keypoint =
      FeatureKeypoint::FromShapeParameters(1, 2, 2, 3, -M_PI / 2, M_PI / 4);
  EXPECT_EQ(keypoint.x, 1.0f);
  EXPECT_EQ(keypoint.y, 2.0f);
  EXPECT_NEAR(keypoint.a11, 0.0f, 1e-6);
  EXPECT_NEAR(keypoint.a12, 2.12132025f, 1e-6);
  EXPECT_NEAR(keypoint.a21, -2.0f, 1e-6);
  EXPECT_NEAR(keypoint.a22, 2.12132025f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeScale(), 2.5f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeScaleX(), 2.0f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeScaleY(), 3.0f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeOrientation(), -M_PI / 2, 1e-6);
  EXPECT_NEAR(keypoint.ComputeShear(), M_PI / 4, 1e-6);

  keypoint =
      FeatureKeypoint::FromShapeParameters(1, 2, 2, 3, M_PI / 2, M_PI / 4);
  EXPECT_EQ(keypoint.x, 1.0f);
  EXPECT_EQ(keypoint.y, 2.0f);
  EXPECT_NEAR(keypoint.a11, 0.0f, 1e-6);
  EXPECT_NEAR(keypoint.a12, -2.12132025f, 1e-6);
  EXPECT_NEAR(keypoint.a21, 2.0f, 1e-6);
  EXPECT_NEAR(keypoint.a22, -2.12132025f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeScale(), 2.5f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeScaleX(), 2.0f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeScaleY(), 3.0f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeOrientation(), M_PI / 2, 1e-6);
  EXPECT_NEAR(keypoint.ComputeShear(), M_PI / 4, 1e-6);

  keypoint.Rescale(2, 2);
  EXPECT_EQ(keypoint.x, 2.0f);
  EXPECT_EQ(keypoint.y, 4.0f);
  EXPECT_NEAR(keypoint.a11, 2 * 0.0f, 1e-6);
  EXPECT_NEAR(keypoint.a12, 2 * -2.12132025f, 1e-6);
  EXPECT_NEAR(keypoint.a21, 2 * 2.0f, 1e-6);
  EXPECT_NEAR(keypoint.a22, 2 * -2.12132025f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeScale(), 2 * 2.5f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeScaleX(), 2 * 2.0f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeScaleY(), 2 * 3.0f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeOrientation(), M_PI / 2, 1e-6);
  EXPECT_NEAR(keypoint.ComputeShear(), M_PI / 4, 1e-6);

  keypoint.Rescale(1, 0.5);
  EXPECT_EQ(keypoint.x, 2.0f);
  EXPECT_EQ(keypoint.y, 2.0f);
  EXPECT_NEAR(keypoint.a11, 0.0f, 1e-6);
  EXPECT_NEAR(keypoint.a12, -2.12132025f, 1e-6);
  EXPECT_NEAR(keypoint.a21, 4.0f, 1e-6);
  EXPECT_NEAR(keypoint.a22, -2.12132025f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeScale(), 3.5f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeScaleX() - 2, 2.0f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeScaleY(), 3.0f, 1e-6);
  EXPECT_NEAR(keypoint.ComputeOrientation(), M_PI / 2, 1e-6);
  EXPECT_NEAR(keypoint.ComputeShear(), M_PI / 4, 1e-6);

  EXPECT_EQ(keypoint, keypoint);
  EXPECT_NE(keypoint, FeatureKeypoint(1, 2, 1, 0));
}

TEST(FeatureKeypoint, Rot90) {
  FeatureKeypoint kp(1.0f, 2.0f, 3.0f, 4.0f, 5.0f, 6.0f);
  int w = 10, h = 20;

  FeatureKeypoint kp1 = kp;
  kp1.Rot90(1, w, h);  // 90 CCW
  EXPECT_FLOAT_EQ(kp1.x, 2.0f);
  EXPECT_FLOAT_EQ(kp1.y, 10.0f - 1.0f);
  EXPECT_FLOAT_EQ(kp1.a11, 5.0f);
  EXPECT_FLOAT_EQ(kp1.a12, 6.0f);
  EXPECT_FLOAT_EQ(kp1.a21, -3.0f);
  EXPECT_FLOAT_EQ(kp1.a22, -4.0f);

  FeatureKeypoint kp2 = kp;
  kp2.Rot90(2, w, h);  // 180 CCW
  EXPECT_FLOAT_EQ(kp2.x, 10.0f - 1.0f);
  EXPECT_FLOAT_EQ(kp2.y, 20.0f - 2.0f);
  EXPECT_FLOAT_EQ(kp2.a11, -3.0f);
  EXPECT_FLOAT_EQ(kp2.a12, -4.0f);
  EXPECT_FLOAT_EQ(kp2.a21, -5.0f);
  EXPECT_FLOAT_EQ(kp2.a22, -6.0f);

  FeatureKeypoint kp3 = kp;
  kp3.Rot90(3, w, h);  // 270 CCW
  EXPECT_FLOAT_EQ(kp3.x, 20.0f - 2.0f);
  EXPECT_FLOAT_EQ(kp3.y, 1.0f);
  EXPECT_FLOAT_EQ(kp3.a11, -5.0f);
  EXPECT_FLOAT_EQ(kp3.a12, -6.0f);
  EXPECT_FLOAT_EQ(kp3.a21, 3.0f);
  EXPECT_FLOAT_EQ(kp3.a22, 4.0f);

  FeatureKeypoint kp_identity = kp;
  kp_identity.Rot90(0, w, h);
  EXPECT_EQ(kp_identity, kp);

  FeatureKeypoint kp_identity4 = kp;
  kp_identity4.Rot90(4, w, h);
  EXPECT_EQ(kp_identity4, kp);

  FeatureKeypoint kp_neg1 = kp;
  kp_neg1.Rot90(-1, w, h);  // same as 3
  EXPECT_EQ(kp_neg1, kp3);
}

TEST(FeatureDescriptors, Nominal) {
  FeatureDescriptors descriptors(FeatureExtractorType::SIFT,
                                 FeatureDescriptorsData::Random(2, 3));
  EXPECT_EQ(descriptors.type, FeatureExtractorType::SIFT);
  EXPECT_EQ(descriptors.data.rows(), 2);
  EXPECT_EQ(descriptors.data.cols(), 3);
  EXPECT_EQ(descriptors.data(0, 0), descriptors.data.data()[0]);
  EXPECT_EQ(descriptors.data(0, 1), descriptors.data.data()[1]);
  EXPECT_EQ(descriptors.data(0, 2), descriptors.data.data()[2]);
  EXPECT_EQ(descriptors.data(1, 0), descriptors.data.data()[3]);
  EXPECT_EQ(descriptors.data(1, 1), descriptors.data.data()[4]);
  EXPECT_EQ(descriptors.data(1, 2), descriptors.data.data()[5]);
}

TEST(FeatureDescriptors, SiftConversion) {
  // SIFT uses value cast (uint8 <-> float)
  const FeatureDescriptors original(FeatureExtractorType::SIFT,
                                    FeatureDescriptorsData::Random(10, 128));
  const FeatureDescriptorsFloat as_float = original.ToFloat();
  EXPECT_EQ(as_float.type, FeatureExtractorType::SIFT);
  EXPECT_EQ(as_float.data.cols(), original.data.cols());
  EXPECT_EQ(as_float.data, original.data.cast<float>());

  const FeatureDescriptors recovered = as_float.ToBytes();
  EXPECT_EQ(recovered.type, FeatureExtractorType::SIFT);
  EXPECT_EQ(recovered.data, original.data);
}

TEST(FeatureDescriptors, AlikedConversion) {
  // ALIKED uses reinterpret cast (float32 bytes <-> float)
  const FeatureDescriptors original(FeatureExtractorType::ALIKED_N16ROT,
                                    FeatureDescriptorsData::Random(10, 512));
  const FeatureDescriptorsFloat as_float = original.ToFloat();
  EXPECT_EQ(as_float.type, FeatureExtractorType::ALIKED_N16ROT);
  EXPECT_EQ(as_float.data.cols() * sizeof(float), original.data.cols());

  const FeatureDescriptors recovered = as_float.ToBytes();
  EXPECT_EQ(recovered.type, FeatureExtractorType::ALIKED_N16ROT);
  EXPECT_EQ(recovered.data, original.data);
}

TEST(FeatureMatches, Nominal) {
  FeatureMatch match;
  EXPECT_EQ(match.point2D_idx1, kInvalidPoint2DIdx);
  EXPECT_EQ(match.point2D_idx2, kInvalidPoint2DIdx);
  FeatureMatches matches(1);
  EXPECT_EQ(matches.size(), 1);
  EXPECT_EQ(matches[0].point2D_idx1, kInvalidPoint2DIdx);
  EXPECT_EQ(matches[0].point2D_idx2, kInvalidPoint2DIdx);

  EXPECT_EQ(match, match);
  EXPECT_NE(match, FeatureMatch(0, 1));
}

TEST(KeypointsMatrixConversion, Roundtrip) {
  FeatureKeypoints keypoints;
  keypoints.emplace_back(1.0f, 2.0f, 3.0f, 0.5f);
  keypoints.emplace_back(4.0f, 5.0f, 1.0f, -0.3f);
  keypoints.emplace_back(10.0f, 20.0f, 0.5f, 0.0f);
  const FeatureKeypoints recovered =
      KeypointsFromMatrix(KeypointsToMatrix(keypoints));
  ASSERT_EQ(recovered.size(), keypoints.size());
  for (size_t i = 0; i < keypoints.size(); ++i) {
    EXPECT_EQ(recovered[i].x, keypoints[i].x);
    EXPECT_EQ(recovered[i].y, keypoints[i].y);
    EXPECT_NEAR(recovered[i].ComputeScale(), keypoints[i].ComputeScale(), 1e-5);
    EXPECT_NEAR(recovered[i].ComputeOrientation(),
                keypoints[i].ComputeOrientation(),
                1e-5);
  }
}

TEST(MatchesMatrixConversion, Roundtrip) {
  FeatureMatches matches;
  matches.emplace_back(0, 5);
  matches.emplace_back(3, 7);
  matches.emplace_back(100, 200);
  EXPECT_EQ(MatchesFromMatrix(MatchesToMatrix(matches)), matches);
}

}  // namespace
}  // namespace colmap

File: src/pycolmap/feature/types_test.py
import math

import numpy as np

import pycolmap


def test_feature_extractor_type_enum():
    assert pycolmap.FeatureExtractorType.UNDEFINED is not None
    assert pycolmap.FeatureExtractorType.SIFT is not None


def test_feature_extractor_type_int_construction():
    sift = pycolmap.FeatureExtractorType.SIFT
    value = int(sift)
    assert isinstance(value, int)


def test_feature_descriptors_default_init():
    descriptors = pycolmap.FeatureDescriptors()
    assert descriptors is not None


def test_feature_descriptors_type_readwrite():
    descriptors = pycolmap.FeatureDescriptors()
    descriptors.type = pycolmap.FeatureExtractorType.SIFT
    assert descriptors.type == pycolmap.FeatureExtractorType.SIFT


def test_feature_descriptors_data_readwrite():
    data = np.array([[1, 2, 3], [4, 5, 6]], dtype=np.uint8)
    descriptors = pycolmap.FeatureDescriptors(
        type=pycolmap.FeatureExtractorType.SIFT, data=data
    )
    np.testing.assert_array_equal(descriptors.data, data)
    new_data = np.array([[10, 20, 30]], dtype=np.uint8)
    descriptors.data = new_data
    np.testing.assert_array_equal(descriptors.data, new_data)


def test_feature_descriptors_float_default_init():
    descriptors = pycolmap.FeatureDescriptorsFloat()
    assert descriptors is not None


def test_feature_descriptors_float_type_readwrite():
    descriptors = pycolmap.FeatureDescriptorsFloat()
    descriptors.type = pycolmap.FeatureExtractorType.SIFT
    assert descriptors.type == pycolmap.FeatureExtractorType.SIFT


def test_feature_descriptors_float_data_readwrite():
    data = np.array([[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]], dtype=np.float32)
    descriptors = pycolmap.FeatureDescriptorsFloat(
        type=pycolmap.FeatureExtractorType.SIFT, data=data
    )
    np.testing.assert_array_almost_equal(descriptors.data, data)
    new_data = np.array([[10.0, 20.0, 30.0]], dtype=np.float32)
    descriptors.data = new_data
    np.testing.assert_array_almost_equal(descriptors.data, new_data)


def test_feature_keypoint_default_init():
    keypoint = pycolmap.FeatureKeypoint()
    assert keypoint is not None


def test_feature_keypoint_xy_readwrite():
    keypoint = pycolmap.FeatureKeypoint()
    keypoint.x = 10.5
    keypoint.y = 20.5
    assert keypoint.x == 10.5
    assert keypoint.y == 20.5


def test_feature_keypoint_affine_readwrite():
    keypoint = pycolmap.FeatureKeypoint()
    keypoint.a11 = 1.0
    keypoint.a12 = 0.5
    keypoint.a21 = -0.5
    keypoint.a22 = 1.0
    assert keypoint.a11 == 1.0
    assert keypoint.a12 == 0.5
    assert keypoint.a21 == -0.5
    assert keypoint.a22 == 1.0


def test_feature_keypoint_compute_scale():
    keypoint = pycolmap.FeatureKeypoint()
    scale = keypoint.compute_scale()
    assert isinstance(scale, float)


def test_feature_keypoint_compute_scale_x():
    keypoint = pycolmap.FeatureKeypoint()
    scale_x = keypoint.compute_scale_x()
    assert isinstance(scale_x, float)


def test_feature_keypoint_compute_scale_y():
    keypoint = pycolmap.FeatureKeypoint()
    scale_y = keypoint.compute_scale_y()
    assert isinstance(scale_y, float)


def test_feature_keypoint_compute_orientation():
    keypoint = pycolmap.FeatureKeypoint()
    orientation = keypoint.compute_orientation()
    assert isinstance(orientation, float)


def test_feature_keypoint_compute_shear():
    keypoint = pycolmap.FeatureKeypoint()
    shear = keypoint.compute_shear()
    assert isinstance(shear, float)


def test_feature_keypoint_rescale():
    keypoint = pycolmap.FeatureKeypoint()
    keypoint.x = 10.0
    keypoint.y = 20.0
    keypoint.a11 = 1.0
    keypoint.a12 = 0.0
    keypoint.a21 = 0.0
    keypoint.a22 = 1.0
    keypoint.rescale(2.0, 2.0)
    assert keypoint.x == 20.0
    assert keypoint.y == 40.0


def test_feature_keypoint_from_shape_parameters():
    keypoint = pycolmap.FeatureKeypoint.from_shape_parameters(
        1.0, 2.0, 2.0, 2.0, math.pi, 0.0
    )
    assert isinstance(keypoint, pycolmap.FeatureKeypoint)
    assert keypoint.x == 1.0
    assert keypoint.y == 2.0


def test_feature_match_default_init():
    match = pycolmap.FeatureMatch()
    assert match is not None


def test_feature_match_readwrite():
    match = pycolmap.FeatureMatch()
    match.point2D_idx1 = 5
    match.point2D_idx2 = 10
    assert match.point2D_idx1 == 5
    assert match.point2D_idx2 == 10


def test_feature_keypoints_vector():
    keypoints = pycolmap.FeatureKeypoints()
    keypoint = pycolmap.FeatureKeypoint()
    keypoint.x = 1.0
    keypoint.y = 2.0
    keypoints.append(keypoint)
    keypoints.append(pycolmap.FeatureKeypoint())
    assert len(keypoints) == 2
    count = 0
    for keypoint in keypoints:
        assert isinstance(keypoint, pycolmap.FeatureKeypoint)
        count += 1
    assert count == 2


def test_feature_matches_vector():
    matches = pycolmap.FeatureMatches()
    matches.append(pycolmap.FeatureMatch(0, 1))
    matches.append(pycolmap.FeatureMatch(2, 3))
    assert len(matches) == 2


def test_keypoints_to_from_matrix_roundtrip():
    keypoints = pycolmap.FeatureKeypoints()
    keypoint = pycolmap.FeatureKeypoint()
    keypoint.x = 10.0
    keypoint.y = 20.0
    keypoint.a11 = 1.0
    keypoint.a12 = 0.0
    keypoint.a21 = 0.0
    keypoint.a22 = 1.0
    keypoints.append(keypoint)
    matrix = pycolmap.keypoints_to_matrix(keypoints)
    assert matrix.shape[0] == 1
    assert matrix.shape[1] == 4
    recovered = pycolmap.keypoints_from_matrix(matrix)
    assert len(recovered) == 1
    assert recovered[0].x == keypoint.x
    assert recovered[0].y == keypoint.y


def test_matches_to_from_matrix_roundtrip():
    matches = pycolmap.FeatureMatches()
    matches.append(pycolmap.FeatureMatch(0, 5))
    matches.append(pycolmap.FeatureMatch(1, 3))
    matrix = pycolmap.matches_to_matrix(matches)
    assert matrix.shape == (2, 2)
    recovered = pycolmap.matches_from_matrix(matrix)
    assert len(recovered) == 2
    assert recovered[0].point2D_idx1 == 0
    assert recovered[0].point2D_idx2 == 5
    assert recovered[1].point2D_idx1 == 1
    assert recovered[1].point2D_idx2 == 3
